### PR TITLE
Fix windows build linkage errors

### PR DESCRIPTION
**Pull request by ziab**
I saw 2 warnings however:
`warning C4996: 'mbstowcs': This function or variable may be unsafe. Consider using mbstowcs_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.`

Could you check if it compiles on GCC now?
Update:
Oh I forgot there is Travis that can indicate if it compiles on GCC.

### CODE REVIEW

**Comment by ziab** [comment]
I see that GCC and Clang now fail with the same error as windows build failed with. I think make files should be modified to compile Lua as C++

File: LuaCppInterface/luacppinclude.h [hunk 1]
```diff
@@ -2,7 +2,7 @@
 #define LUACPPINCLUDE
 
 
-extern "C" {
+extern "C++" {
 #include "lua.h"
```
Comment (davidsiaw) [inline LuaCppInterface/luacppinclude.h]: this causes the headers to be compiled as C++ code, mangling their names. the Lua C lib built in C will have different names. Perhaps you need to compile your Lua code on windows in as C code.

**Comment by davidsiaw** [comment]
No. Lua is a C library and should be compiled thus. Please change the windows build to compile the Lua sources as C code and revert the extern "C++" back to extern "C"

**Comment by ziab** [comment]
Yeah, it's definitely a C library, but you can compile it using C++ compiler and get a bit different behavior.

The implementation of `luaL_error()` uses `LUAI_THROW`, which definition is:
```
/*
** LUAI_THROW/LUAI_TRY define how Lua does exception handling. By
** default, Lua handles errors with exceptions when compiling as
** C++ code, with _longjmp/_setjmp when asked to use them, and with
** longjmp/setjmp otherwise.
*/
#if !defined(LUAI_THROW)

#if defined(__cplusplus) && !defined(LUA_USE_LONGJMP)
/* C++ exceptions */
#define LUAI_THROW(L,c)		throw(c)
#define LUAI_TRY(L,c,a) \
	try { a } catch(...) { if ((c)->status == 0) (c)->status = -1; }
#define luai_jmpbuf		int  /* dummy variable */

#elif defined(LUA_USE_ULONGJMP)
/* in Unix, try _longjmp/_setjmp (more efficient) */
#define LUAI_THROW(L,c)		_longjmp((c)->b, 1)
#define LUAI_TRY(L,c,a)		if (_setjmp((c)->b) == 0) { a }
#define luai_jmpbuf		jmp_buf

#else
/* default handling with long jumps */
#define LUAI_THROW(L,c)		longjmp((c)->b, 1)
#define LUAI_TRY(L,c,a)		if (setjmp((c)->b) == 0) { a }
#define luai_jmpbuf		jmp_buf

#endif
```
Long jumps may lead to memory leaks. If it's compiled as C++ the stack will be unwound properly and object destructors will be called.

**Comment by davidsiaw** [comment]
Interesting. This actually means that it would be a good idea to compile Lua into the lib itself. This might break expectations. This change might warrant a major version change

**Comment by ziab** [comment]
It probably makes sense to compile Lua into the lib, this is the way I do it my project actually. But there may be users who already have Lua library and some legacy code written and want to use LuaCppInterface. Compiling Lua as C++ should be suggested in such case. This way of compilation can be set as default in distributed .proj files and make files.

I just thought if you put `extern "C++"` it would cause linkage errors if Lua compiled as C, which is rather good. These errors just hint that LuaCppInterface is supposed to be used with Lua compiled as C++.

**Comment by davidsiaw** [comment]
Let us leave it compiling as C for now. All the projects (Mac OS X, iOS, Android, Linux) will need to be changed to compile Lua as C++. That can be released as a newer version.

**Comment by ziab** [comment]
Sure, I've updated MSVC projects. All the checks seem to pass now.

David, could you review this code:

```
	static int staticFunction(lua_State* state)
	{
		std::function<void(T1)>* func = (std::function<void(T1)>*)lua_touserdata(state, lua_upvalueindex(1));
		lua_remove(state, 1);

		if (lua_gettop(state) < parameters)
		{
			return luaL_error(state, "expected parameters count: %d", parameters);
		}

		try
		{
			arg1type arg1 = popper<arg1type>::pop(LuaNoDestructor(state));
			func->operator()(arg1);
		}
		catch (LuaNil& e)
		{
			UNREFERENCED_PARAMETER(e);
		}
		return 0;
	}
```

It seems like even if `luaL_error` long jumps (because compiled as C) there should be no memory leak since no destructible objects are allocated on the stack, no memory on the heap allocated in those functions either and `std::function<void(T1)>* func` is just a raw pointer. So it's OK to compile as C for now.

**Pull request merged**